libnvme/src/firmware.rs

```diff
@@ -0,0 +1,306 @@
+pub enum NvmeSlotError {
+    #[error("NVMe slots must be between 1 and 7 but got {0}")]
+    InvalidSlotNumber(usize),
+    #[error("NVMe device does not have slot {0}")]
+    DoesNotExisit(u32),
+    #[error("libnvme error: {0}")]
+    NvmeError(#[from] NvmeError),
+}
+
+// NOTE: we represent this is a u32 because `nvme_fw_commit_req_set_slot` takes
+// a uint32_t as the spec may change in the future to add additional slots.
+/// NVMe Slot Number.
+pub struct NvmeSlot(u32);
+
+// Rust's default interger type is `i32`.
+impl TryFrom<i32> for NvmeSlot {
+    type Error = NvmeSlotError;
+
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        match value {
+            1..=7 => Ok(NvmeSlot(value as u32)),
+            invalid => Err(NvmeSlotError::InvalidSlotNumber(invalid as _)),
+        }
+    }
+}
+
+// Setting a slot via `nvme_fw_commit_req_set_slot` uses a `u32`.
+impl TryFrom<u32> for NvmeSlot {
+    type Error = NvmeSlotError;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            1..=7 => Ok(NvmeSlot(value)),
+            invalid => Err(NvmeSlotError::InvalidSlotNumber(invalid as _)),
+        }
+    }
+}
+
+// The NVMe spec currently uses a u8 for active and next active slots.
+impl TryFrom<u8> for NvmeSlot {
+    type Error = NvmeSlotError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            1..=7 => Ok(NvmeSlot(value as _)),
+            invalid => Err(NvmeSlotError::InvalidSlotNumber(invalid as _)),
+        }
+    }
+}
+
+#[derive(Debug)]
+/// The Firmwware log page for an NVMe controller.
+pub struct FirmwareLogPage {
+    /// The currently active firmware slot.
+    pub active_slot: u8,
+    /// The next slot that will be active on a controller reset.
+    pub next_active_slot: Option<u8>,
+    /// Slot 1 is read-only.
+    pub slot1_is_read_only: bool,
+    /// The number of firmware slots the controller has.
+    pub number_of_slots: usize,
+    firmware_slots: Vec<Option<String>>,
+}
+
+impl FirmwareLogPage {
+    fn init(
+        identify: &ControllerInfoIdentify<'_>,
+        logpage: nvme_fwslot_log_t,
+    ) -> Self {
+        let active_slot = logpage.bitfield1.fw_afi();
+        let slot1_is_read_only =
+            unsafe { (*identify.inner).id_frmw.fw_readonly() };
+
+        // NVMe Spec: "If this field is 0h, then the controller does not
+        // indicate the firmware slot that is going to be activated at the next
+        // Controller Level Reset."
+        let next_active_slot = match logpage.bitfield1.fw_next() {
+            0 => None,
+            slot => Some(slot.saturating_sub(1)),
+        };
+
+        let number_of_slots =
+            unsafe { (*identify.inner).id_frmw.fw_nslot() } as usize;
+        let mut firmware_slots = Vec::with_capacity(number_of_slots);
+        for slot in &logpage.fw_frs[..number_of_slots] {
+            // FIXME there's probably a better way to do this? Basically the
```

Comment:
I think you can skip the `Vec` and `CString` altogether. Untested, but something like:

```rust
            // Convert the &[c_char] slot to a &[u8]
            let u8slot = unsafe {
                slice::from_raw_parts(slot.as_ptr().cast::<u8>(), slot.len())
            };
            if u8slot[0] == b'\0' {
                firmware_slots.push(None);
                continue;
            }
            let firmware = String::from_utf8_lossy(u8slot).to_string();
            firmware_slots.push(Some(firmware));
```

Reply:
Fixed in dff4996b